parser2/src/lib.rs

```diff
@@ -0,0 +1,200 @@
+pub fn parse_type(src: &str) -> Result<Type, Error> {
+    grammar::parse_Type(src, Tokenizer::new(src))
+}
+
+// Patterns
+
+/// The representation of gluon's pattern syntax
+#[derive(Clone, PartialEq, Debug)]
+pub enum Pattern {
+    Ident(Ident),
+    Ctor(Ident, Vec<Pattern>),
+    Record(Vec<(Ident, Pattern)>),
+}
+
+impl Pattern {
+    /// Construct a tuple, desugaring it to a record pattern
+    pub fn tuple<Elems>(elems: Elems) -> Pattern
+        where Elems: IntoIterator<Item = Pattern>
+    {
+        Pattern::Record(desugar_tuple(elems))
+    }
+}
+
+// Expressions
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Literal {
+    Byte(u8),
+    Int(i64),
+    Float(f64),
+    String(String),
+    Char(char),
+}
+
+pub fn parse_literal(src: &str) -> Result<Literal, Error> {
+    grammar::parse_Literal(src, Tokenizer::new(src))
+}
+
+/// Arm of a match expression (see `Expr::Match`)
+#[derive(Clone, PartialEq, Debug)]
+pub struct MatchArm {
+    pattern: Pattern,
+    body: Box<Expr>,
+}
+
+impl MatchArm {
+    pub fn new(pattern: Pattern, body: Expr) -> MatchArm {
+        MatchArm {
+            pattern: pattern,
+            body: Box::new(body),
+        }
+    }
+}
+
+/// Binding in a let expression (see `Expr::LetBindings`)
+#[derive(Clone, PartialEq, Debug)]
+pub struct ValueBinding {
+    pattern: Pattern,
+    value: Expr,
+}
+
+impl ValueBinding {
+    pub fn new(pattern: Pattern, value: Expr) -> ValueBinding {
+        ValueBinding {
+            pattern: pattern,
+            value: value,
+        }
+    }
+}
+
+/// The representation of gluon's expression syntax
+#[derive(Clone, PartialEq, Debug)]
+pub enum Expr {
+    /// Identifiers
+    Ident(Ident),
+    /// Literal values
+    Literal(Literal),
+    /// Field getter functions
+    Getter(Ident),
+    /// Records
+    Record(Vec<(Ident, Expr)>),
+    /// Function application, eg. `f x`
+    App(Box<Expr>, Vec<Expr>),
+    /// Field projection, eg. `value.field`
+    Project(Box<Expr>, Ident),
+    /// Binary infix operator application, eg. `f >> g`
+    Infix(Box<Expr>, String, Box<Expr>),
+    /// Lambda abstraction, eg. `\x y -> x * y`
+    Lambda(Vec<Pattern>, Box<Expr>),
+    /// If-then-else conditional
+    IfElse(Box<Expr>, Box<Expr>, Box<Expr>),
+    /// Pattern match expression
+    Match(Box<Expr>, Vec<MatchArm>),
+    /// Multiple value bindings
+    LetBindings(Vec<ValueBinding>, Box<Expr>),
+    /// Declare a series of type aliases (should eventually be desugared into a let binding)
+    TypeBindings(Vec<TypeBinding>, Box<Expr>),
+    /// A block of sequenced (usually procedural) expressions.
+    Block(Vec<Expr>),
+}
+
+impl Expr {
+    /// Construct a tuple, desugaring it to a record expression
+    pub fn tuple<Elems>(elems: Elems) -> Expr
+        where Elems: IntoIterator<Item = Expr>
+    {
+        Expr::Record(desugar_tuple(elems))
+    }
+}
+
+pub fn parse_expr(src: &str) -> Result<Expr, Error> {
+    grammar::parse_Expr(src, Tokenizer::new(src))
+}
```

Comment:
Why have you defined a new AST instead of using (and modifying the existing on)?